src/tardev-snapshotter/docs/ROADMAP.md

```diff
@@ -0,0 +1,149 @@
+# RFC: Code Integrity for OCI/containerd Containers using erofs-snapshotter
+
+This document outlines the strategic roadmap for enhancing code integrity in OCI
+Containers through erofs-snapshotter. The purpose of this document to seek
+feedback from the community and align on the path forward for implementing the
+outlined milestones.
+
+## Objective
+
+The primary objective is to ensure that container runtime integrity is enforced
+at the kernel level using
+[dm-verity](https://www.kernel.org/doc/html/latest/admin-guide/device-mapper/verity.html).
+Additionally, only binaries originating from the original container image layers
+should be permitted to execute, enforced via
+[IPE](https://docs.kernel.org/next/admin-guide/LSM/ipe.html).
+
+To achieve this, each container image layer will be backed by a dm-verity
+device. The signed root hash of the corresponding Merkle tree will be provided
+to the kernel, enabling provenance validation through kernel keyrings.
+Subsequently, an IPE policy can be deployed on the host to restrict execution to
+binaries residing on these signed dm-verity block devices.
+
+This approach extends the current containerd model, which verifies artifact
+digests during registry downloads but does not enforce artifact integrity
+post-pull. By leveraging IPE, we further ensure that only trusted code is
+executed.
+
+Our aim is to deliver these capabilities seamlessly, minimizing the need for
+modifications to existing build and runtime environments by integrating with OCI
+registries.
+
+The following sections detail the milestones required to realize these
+objectives.
+
+## Milestones
+
+### Milestone 1: Foundational Integration with containerd/erofs-snapshotter
+
+This milestone establishes the essential components for code integrity in OCI
+Containers, including support for tar index, dm-verity, and dm-verity signatures
+within the erofs-snapshotter.
+
+Key deliverables:
+
+1. **Tar Index Mode for EROFS Conversion** Introduce an alternative method for
+   converting image layer tarballs to EROFS within the snapshotter, utilizing
+   tar index mode.
+
+2. **dm-verity Support Finalization** Complete dm-verity integration in the
```

Comment:
If people need to associate keys anyway, why not require these artifacts to be in the registry as well and skip on-the-fly generation?

Reply:
From our perspective, we only care about the registry based solution. But the erofs-snapshotter already has logic for on the node generation of both the erofs blobs and the dm-verity merkle tree, and so instead of keeping that as a separate flow, it felt nice to be able to integrate it as a fallback mechanism.